### PR TITLE
Combined save lowest source and target epoch inside save attestation history

### PR DESCRIPTION
**Pull request by shayzluf**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**
for better validator performance it is good to reduce lock times on db by combining the data save operations

**Before 1.5s**
Screen Shot 2020-12-10 at 7 31 14
**After 0.43s**
Screen Shot 2020-12-10 at 7 37 12

### CODE REVIEW

File: validator/db/kv/attestation_history_v2.go [hunk 1]
```diff
@@ -45,7 +47,7 @@ func (store *Store) AttestationHistoryForPubKeyV2(ctx context.Context, publicKey
 		bucket := tx.Bucket(newHistoricAttestationsBucket)
 		enc := bucket.Get(publicKey[:])
 		if len(enc) == 0 {
-			attestationHistory = NewAttestationHistoryArray(0)
+			attestationHistory = NewAttestationHistoryArray(53999)
```
Comment (shayzluf) [inline validator/db/kv/attestation_history_v2.go:50]: this is a tweak for getting the full size array in interop on slot 1.
have to remove !!!!

File: validator/db/kv/attestation_history_v2.go [hunk 2]
```diff
@@ -59,19 +61,42 @@ func (store *Store) AttestationHistoryForPubKeyV2(ctx context.Context, publicKey
...
+	lowestSourceEpoch,
+	lowestTargetEpoch uint64,
+) error {
 	ctx, span := trace.StartSpan(ctx, "Validator.SaveAttestationHistoryForPubKeyV2")
 	defer span.End()
-	err := store.update(func(tx *bolt.Tx) error {
-		bucket := tx.Bucket(newHistoricAttestationsBucket)
-		return bucket.Put(pubKey[:], history)
-	})
+	go func() {
```
Comment (shayzluf) [inline validator/db/kv/attestation_history_v2.go:73]: this is an experiment. seems to work better but still if io cant handle the load it just postpones the issue

File: validator/db/kv/attestation_history_v2_test.go [hunk 3]
```diff
@@ -49,14 +50,82 @@ func TestStore_AttestedPublicKeys(t *testing.T) {
...
+		require.NoError(t, validatorDB.ClearDB(), "Failed to clear database")
+	})
+	p0 := [48]byte{0}
+	p1 := [48]byte{1}
+	// Can save.
+	err = validatorDB.SaveAttestationHistoryForPubKeyV2(ctx, p0, NewAttestationHistoryArray(0), 100, 100)
+	require.NoError(t, err)
+	err = validatorDB.SaveAttestationHistoryForPubKeyV2(ctx, p1, NewAttestationHistoryArray(0), 200, 200)
+	require.NoError(t, err)
+	time.Sleep(50 * time.Millisecond)
```
Comment (shayzluf) [inline validator/db/kv/attestation_history_v2_test.go:76]: part of the experiment

File: validator/db/kv/attestation_history_v2_test.go [hunk 4]
```diff
@@ -49,14 +50,82 @@ func TestStore_AttestedPublicKeys(t *testing.T) {
...
+	got, err = validatorDB.LowestSignedTargetEpoch(ctx, p1)
+	require.NoError(t, err)
+	require.Equal(t, uint64(200), got)
+
+	// Can replace.
+	err = validatorDB.SaveAttestationHistoryForPubKeyV2(ctx, p0, NewAttestationHistoryArray(0), 99, 99)
+	require.NoError(t, err)
+	err = validatorDB.SaveAttestationHistoryForPubKeyV2(ctx, p1, NewAttestationHistoryArray(0), 199, 199)
+	require.NoError(t, err)
+	time.Sleep(50 * time.Millisecond)
```
Comment (shayzluf) [inline validator/db/kv/attestation_history_v2_test.go:95]: same ^

File: validator/db/kv/attestation_history_v2_test.go [hunk 5]
```diff
@@ -49,14 +50,82 @@ func TestStore_AttestedPublicKeys(t *testing.T) {
...
+	got, err = validatorDB.LowestSignedTargetEpoch(ctx, p1)
+	require.NoError(t, err)
+	require.Equal(t, uint64(199), got)
+
+	// Can not replace.
+	err = validatorDB.SaveAttestationHistoryForPubKeyV2(ctx, p0, NewAttestationHistoryArray(0), 100, 100)
+	require.NoError(t, err)
+	err = validatorDB.SaveAttestationHistoryForPubKeyV2(ctx, p1, NewAttestationHistoryArray(0), 200, 200)
+	require.NoError(t, err)
+	time.Sleep(50 * time.Millisecond)
```
Comment (shayzluf) [inline validator/db/kv/attestation_history_v2_test.go:114]: same ^

**Comment by rauljordan** [comment]
@shayzluf now that this is in develop, can we extract the core logic of combining them into a separate PR? thanks for doing this

**Comment by shayzluf** [comment]
@rauljordan checking now

**Pull request closed without merging**